gluster/subvol/cmd/gluster-subvol-provisioner/gluster-subvol-provisioner.go

```diff
@@ -0,0 +1,606 @@
+	glog.Infof("going to mount supervol %s for PV %s: %s", supervol, pv.Name, mountCmd)
+
+	// os.StartProcess() does not allow deamons (fuse) to continue, using
+	// exec.Command instead.
+	cmd := exec.Command(mountCmd[0], mountCmd[1:]...)
+	err = cmd.Start()
+	if err != nil {
+		glog.Errorf("failed to mount supervol %s: %s", supervol, err)
+		return "", err
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		return "", fmt.Errorf("failed to mount supervol %s, process exited with %v", supervol, err)
+	}
+
+	p.mtab[pv.Name] = mountpoint
+
+	return mountpoint, nil
+}
+
+func (p *glusterSubvolProvisioner) makeSubvolPath(mountpoint, pvcNS string, pvcUID types.UID) string {
+	return fmt.Sprintf("%s/%s/pvc-%s", mountpoint, pvcNS, pvcUID)
+}
+
+func (p *glusterSubvolProvisioner) makeMountPath(supervolPV *v1.PersistentVolume, pvcNS string, pvcUID types.UID) string {
+	return fmt.Sprintf("%s/%s/pvc-%s", supervolPV.Spec.Glusterfs.Path, pvcNS, pvcUID)
+}
+
+func (p *glusterSubvolProvisioner) copyEndpoints(sourceNS string, sourcePV *v1.PersistentVolume, destNS string, destPVCName string) (*v1.Endpoints, error) {
+	// Need to copy the endpoints from the supervol to the new PVC. A
+	// reference of the endpoints name is not sufficient, it can be in an
+	// other namespace.
+	sourceEP, err := p.client.CoreV1().Endpoints(sourceNS).Get(sourcePV.Spec.Glusterfs.EndpointsName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("endpoint %s for source PV %s can not be found: %s", sourcePV.Spec.Glusterfs.EndpointsName, sourcePV.Name, err)
+	}
+
+	ep := sourceEP.DeepCopy()
+	ep.ObjectMeta = metav1.ObjectMeta{
+		Namespace: destNS,
+		Name:      dynamicEpSvcPrefix + destPVCName,
+		Labels: map[string]string{
+			"gluster.org/provisioned-for-pvc": destPVCName,
+		},
+	}
+
+	_, err = p.client.CoreV1().Endpoints(destNS).Create(ep)
+	if err != nil && errors.IsAlreadyExists(err) {
+		glog.V(1).Infof("endpoint %s already exist in namespace %s, that is ok", destPVCName, destNS)
+		err = nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to create endpoint %s: %s", ep.ObjectMeta.Name, err)
+	}
+
+	return ep, nil
+}
+
+// Pre-check for the requirements of the PVC-request and the supervol PVC.
+// - certain features of the PVC-request may not be supported
+// - the supervol PVC has some requirements that it needs to fulfil
+func (p *glusterSubvolProvisioner) validateProvisionRequirements(pvcReq, supervolPVC *v1.PersistentVolumeClaim) error {
+	if pvcReq.Spec.Selector != nil {
+		return fmt.Errorf("claim Selector is not supported")
+	}
+
+	if !util.AccessModesContainedInAll(accessModes, pvcReq.Spec.AccessModes) {
+		return fmt.Errorf("invalid AccessModes %v: only AccessModes %v are supported", pvcReq.Spec.AccessModes, accessModes)
+	}
+
+	// check permission mode, the PV should be RWX as it is mounted here,
+	// and we'll create subdirs as new PVCs
+	if !util.AccessModesContains(supervolPVC.Spec.AccessModes, v1.ReadWriteMany) {
```

Comment:
Check for `RWO` as well.

Reply:
`RWO` is not valid, this provisioner needs to mount the supervol, and the pod will cause mounting of the subvol.